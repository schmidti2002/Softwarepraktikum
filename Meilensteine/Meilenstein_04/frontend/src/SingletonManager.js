// Alles in dieser Klasse könnte statisch sein, ist es aber absichtlich nicht, um die Übergabe
// einer Instanz in den Konstruktor einer Klasse zu "erzwingen", was es einfacher macht, die
// Singletons in Tests zu mocken

export default class SingletonManager {
  #instances = {};

  #constructors = {};

  registerConstructor(name, constructor, override = false) {
    if (this.#constructors[name] === undefined || override) {
      this.#constructors[name] = constructor;
      this.#instances[name] = undefined;
    }
  }

  get(name) {
    if (this.#instances[name] !== undefined) {
      return this.#instances[name];
    }
    if (this.#constructors[name] === undefined) {
      return undefined;
    }
    const instance = this.#constructors[name]();
    this.#instances[name] = instance;
    return instance;
  }
}
