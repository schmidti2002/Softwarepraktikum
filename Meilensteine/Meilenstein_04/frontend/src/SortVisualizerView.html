<style>
    #sortview-container {
        display: flex;
        justify-content: flex-start;
        align-items: flex-end;
        height: 510px;
        /* border: 1px solid #ccc; */
        margin-top: 20px;
        overflow: hidden;
    }

    .bar {
        background-color: #00ba77;
        transition: height 0.3s;
        display: flex;
        flex-direction: column;
        align-items: center;
        margin-right: 2px; /* Dinstanz zwischen Balken */
        position: relative;
    }

    .bar span {
        margin-top: 5px;
        position: absolute;
        bottom: 0;
    }
</style>
<div id="sortview-container"></div>