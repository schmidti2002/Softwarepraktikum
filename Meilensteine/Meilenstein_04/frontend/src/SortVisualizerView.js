import VisualizerView from './VisualizerView';

// Visualisiert ein Array von Zahlen als Säulendiagram
export default class SortVisualizerView extends VisualizerView {
  constructor(parentNode, eventReporter) {
    super('SortVisualizerView', parentNode, eventReporter);
    this.initPromise.then(() => {
      this.container = document.getElementById('sortview-container');
    });
  }

  renderData(data) {
    this.container.innerHTML = '';

    for (let i = 0; i < data.length; i++) { // For erstellt alle Bars
      const bar = document.createElement('div');
      bar.className = 'bar';
      bar.style.width = '20px'; // Skaliere Breite der Bars

      // Skaliere die Höhe der Bars:
      bar.style.height = `${data[i] * 10 + 10}px`;
      bar.innerHTML = `<span>${data[i]}</span>`;
      this.container.appendChild(bar);
    }
  }
}
