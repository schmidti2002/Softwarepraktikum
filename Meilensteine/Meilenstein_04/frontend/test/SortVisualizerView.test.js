import {
  describe, expect, it, beforeAll,
} from '@jest/globals';
import SortVisualizerView from '../src/SortVisualizerView';
import SingletonManager from '../src/SingletonManager';
import { mockFetchHtml, awaitAllAsync } from './testUtils.test';

describe('SortVisualizerView.test.js', () => {
  mockFetchHtml('../src/SortVisualizerView.html');
  let sortVisualizerView;
  beforeAll(() => {
    // Mock the HTML structure expected by the component
    document.body.innerHTML = '<div id="sortview-container"></div>';

    // Create an instance of SortVisualizerView
    sortVisualizerView = new SortVisualizerView(document.body, (new SingletonManager()).get('EventReporter'));

    // Return the initPromise to ensure it resolves before tests
    return sortVisualizerView.initPromise;
  });

  it.each([
    [0, 1, 2, 3, 4, 5],
    [50, 10, 42],
    [1, 5, 13, 20],
  ])('renderData test', async (data) => {
    sortVisualizerView.renderData(data);
    await awaitAllAsync();
    expect(sortVisualizerView.container.parentElement.innerHTML).toMatchSnapshot();
  });
});
