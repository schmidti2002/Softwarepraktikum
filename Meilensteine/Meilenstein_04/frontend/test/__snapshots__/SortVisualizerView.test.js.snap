// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortVisualizerView.test.js renderData test 1`] = `"<div id="sortview-container"><div class="bar" style="width: 20px; height: 10px;"><span>0</span></div><div class="bar" style="width: 20px; height: 20px;"><span>1</span></div><div class="bar" style="width: 20px; height: 30px;"><span>2</span></div><div class="bar" style="width: 20px; height: 40px;"><span>3</span></div><div class="bar" style="width: 20px; height: 50px;"><span>4</span></div><div class="bar" style="width: 20px; height: 60px;"><span>5</span></div></div>"`;

exports[`SortVisualizerView.test.js renderData test 2`] = `"<div id="sortview-container"><div class="bar" style="width: 20px; height: 510px;"><span>50</span></div><div class="bar" style="width: 20px; height: 110px;"><span>10</span></div><div class="bar" style="width: 20px; height: 430px;"><span>42</span></div></div>"`;

exports[`SortVisualizerView.test.js renderData test 3`] = `"<div id="sortview-container"><div class="bar" style="width: 20px; height: 20px;"><span>1</span></div><div class="bar" style="width: 20px; height: 60px;"><span>5</span></div><div class="bar" style="width: 20px; height: 140px;"><span>13</span></div><div class="bar" style="width: 20px; height: 210px;"><span>20</span></div></div>"`;
